### PR TITLE
Write logs to the file

### PR DESCRIPTION
**Pull request by RohovAlex**
Writing logs to the file was added.
Closes: https://github.com/metatech-university/NodeJS-Fastify/issues/4

### CODE REVIEW

**Comment by RohovAlex** [comment]
Sorry, I forgot to create separate branches for these issues

File: main.js [hunk 1]
```diff
@@ -10,9 +11,21 @@ const ws = require('./src/system/ws');
 
 const { loadApplication } = require('./src/system/loader.js');
 const APPLICATION_PATH = path.join(process.cwd(), '../NodeJS-Application');
+const LOG_FOLDER_PATH = './log';
+const date = new Date().toISOString().substring(0, 10);
+const filePath = path.join(LOG_FOLDER_PATH, `${date}.log`);
+
+const streams = [
+  { stream: process.stdout },
+  { stream: pinoLogger.destination(filePath) }
```
Comment (tshemsedinov) [inline main.js:20]:
```suggestion
  { stream: pinoLogger.destination(LOG_FILE) },
```

File: main.js [hunk 2]
```diff
@@ -10,9 +11,21 @@ const ws = require('./src/system/ws');
...
+];
+const multiLogger = pinoLogger(
+  { level: 'info' },
+  pinoLogger.multistream(streams)
```
Comment (tshemsedinov) [inline main.js:24]:
```suggestion
  pinoLogger.multistream(streams),
```

File: main.js [hunk 3]
```diff
@@ -10,9 +11,21 @@ const ws = require('./src/system/ws');
 
 const { loadApplication } = require('./src/system/loader.js');
 const APPLICATION_PATH = path.join(process.cwd(), '../NodeJS-Application');
+const LOG_FOLDER_PATH = './log';
+const date = new Date().toISOString().substring(0, 10);
+const filePath = path.join(LOG_FOLDER_PATH, `${date}.log`);
```
Comment (tshemsedinov) [inline main.js:16]:
```suggestion
const date = new Date().toISOString().substring(0, 10);
const LOG_FILE = path.join('./log', `${date}.log`);
```

**Review by tshemsedinov — CHANGES_REQUESTED** [review]
Just a few naming and style fixes

File: main.js (hunk 3 above)
Comment (tshemsedinov) [reply main.js:16]: `filePath` is ok naming to be inside class constructor, method or just a function but it is not precise enough to be in file-wide context

**Comment by tshemsedinov** [comment]
It will be good to add following tag in commit description

`Closes: https://github.com/metatech-university/NodeJS-Fastify/issues/4`

**Review by nechaido — CHANGES_REQUESTED** [review]
This would be an invalid implementation.
Ont the next day it will continue to write logs to an outdated file.

What you need to do is to:
1. implement a custom writable stream and pass it into `logger: { stream }` option, refer to fastify/pino documentation on the matter.
2. on write dynamically check if current stream is up to date
  2.1 If so -- write to the stream
  2.2 If not -- close an old one, and open a new one
3. Consider sticking with the original approach of having separate files for each log level.

File: main.js [hunk 4]
```diff
@@ -1,6 +1,7 @@
 'use strict';
 
 const fastify = require('fastify');
+const pinoLogger = require('pino');
```
Comment (nechaido) [inline main.js:4]: There is no need to require pino here.
* Either pass custom stream to fastify
* Or create custom pino logger wrapper in /src/logger and pass it to both fastify and sandbox

**Comment by tshemsedinov** [comment]
Hint: I'd advise to use `git rebase` instead of merge. That's about https://github.com/metatech-university/NodeJS-Fastify/pull/7/commits/f5cf16ddec5b75fc7bfbad22b3b2a014dd044e29

**Comment by JekRock** [comment]
As an inspiration:
- [Rotating file stream](https://github.com/iccicci/rotating-file-stream)
- [pino transport for log rotation](https://github.com/ExpediaGroup/pino-rotating-file)

But I personally like the approach [suggested by pino](https://getpino.io/#/docs/help?id=log-rotation) - using external tools for log rotation, like [logrotate](https://linux.die.net/man/8/logrotate).

File: main.js (hunk 1 above)
Comment (RohovAlex) [reply main.js]: Done

File: main.js (hunk 2 above)
Comment (RohovAlex) [reply main.js]: Done

File: main.js (hunk 3 above)
Comment (RohovAlex) [reply main.js]: Done

File: main.js (hunk 4 above)
Comment (RohovAlex) [reply main.js]: Done

**Review by tshemsedinov — APPROVED** [review]
@nechaido I think we can land this

File: main.js [hunk 5]
```diff
@@ -3,18 +3,39 @@
...
+
+const date = new Date().toISOString().substring(0, 10);
+const LOG_FILE = path.join(LOG_FOLDER_PATH, `${date}.log`);
+const logFileStream = fs.createWriteStream(LOG_FILE, { flags: 'a' });
+
+const streamForLogger = {
+  write(msg) {
+    process.stdout.write(msg);
+    const currentDate = new Date().toISOString().substring(0, 10);
+    if (currentDate !== this.date) {
```
Comment (nechaido) [inline main.js:24]: If you rely on using this -- I'd prefer to extract this to a Class.
In any case -- this needs to be extracted into /src/logger so as not to clutter the main entry point.

**Review by nechaido — CHANGES_REQUESTED** [review]

File: main.js (hunk 5 above)
Comment (RohovAlex) [reply main.js]: @nechaido Done

File: main.js [hunk 6]
```diff
@@ -3,18 +3,24 @@
...
-
 const { loadApplication } = require('./src/loader.js');
+
 const APPLICATION_PATH = path.join(process.cwd(), '../NodeJS-Application');
+const LOG_FOLDER_PATH = './log';
+
+const date = new Date().toISOString().substring(0, 10);
+const LOG_FILE = path.join(LOG_FOLDER_PATH, `${date}.log`);
+const logFileStream = fs.createWriteStream(LOG_FILE, { flags: 'a' });
+const streamForLogger = new StreamForLogger(logFileStream, date);
```
Comment (nechaido) [inline main.js:19]: This has nothing to do with the entry point, extract it to the log file.

Moreover, the constructor itself should be responsible for opening the file.

File: main.js (hunk 5 above)
Comment (nechaido) [reply main.js]: Not quite yet, see https://github.com/metatech-university/NodeJS-Fastify/pull/7/files#r1109789506.

File: main.js (hunk 6 above)
Comment (RohovAlex) [reply main.js]: @nechaido Done, please review.

File: main.js [hunk 7]
```diff
@@ -4,17 +4,18 @@ const fastify = require('fastify');
...
-const { Logger } = require('./src/logger.js');
+const { Logger, StreamForLogger } = require('./src/logger.js');
 const http = require('./src/http.js');
 const ws = require('./src/ws.js');
-
 const { loadApplication } = require('./src/loader.js');
+
 const APPLICATION_PATH = path.join(process.cwd(), '../NodeJS-Application');
+const LOG_FOLDER_PATH = './log';
+const streamForLogger = new StreamForLogger(LOG_FOLDER_PATH);
```
Comment (nechaido) [inline main.js:14]: Move this to line 16.

File: src/logger.js [hunk 8]
```diff
@@ -38,6 +40,30 @@ class Logger {
   }
 }
 
+class StreamForLogger {
+  constructor(folderPath) {
+    this.folderPath = folderPath;
+    this.date = new Date().toISOString().substring(0, 10); 
+    this.#createFileStream();
+  }
```
Comment (nechaido) [inline src/logger.js:48]:
```suggestion
  }

```

File: src/logger.js [hunk 9]
```diff
@@ -38,6 +40,30 @@ class Logger {
...
+    const currentDate = new Date().toISOString().substring(0, 10);
+    if (currentDate !== this.date) {
+      this.date = currentDate;
+      this.logFileStream.end();
+      this.#createFileStream();
+    }
+    this.logFileStream.write(msg);
+  }
+
+  #createFileStream() {
```
Comment (nechaido) [inline src/logger.js:60]:
```suggestion
  #createFileStream() {
    if (this.logFileStream) {
      // close existing one
    }
```

File: src/logger.js [hunk 10]
```diff
@@ -38,6 +40,30 @@ class Logger {
...
+    const filePath = path.join(this.folderPath, `${this.date}.log`);
+    this.logFileStream = fs.createWriteStream(filePath, { flags: 'a' });
```
Comment (nechaido) [inline src/logger.js:62]: Make the property private and be sure to always define all the properties of a class in the constructor.

Comment (nechaido) [reply src/logger.js:62]: Or Prop list.

File: main.js (hunk 7 above)
Comment (RohovAlex) [reply main.js:14]: @nechaido do you mean move rows 13, 14 inside IIFE?

Comment (RohovAlex) [reply main.js:14]: Done

File: src/logger.js (hunk 8 above)
Comment (RohovAlex) [reply src/logger.js:48]: Done

File: src/logger.js (hunk 9 above)
Comment (RohovAlex) [reply src/logger.js:60]: Done

File: src/logger.js (hunk 10 above)
Comment (RohovAlex) [reply src/logger.js:62]: Done

**Comment by RohovAlex** [comment]
@nechaido Done, please review.

**Comment by RohovAlex** [comment]
@tshemsedinov  @nechaido please review

File: src/logger.js [hunk 11]
```diff
@@ -38,6 +40,35 @@ class Logger {
...
+    const currentDate = new Date().toISOString().substring(0, 10);
+    if (currentDate !== this.date) {
+      this.date = currentDate;
+      this.#logFileStream = this.#createFileStream();
+    }
+    this.#logFileStream.write(msg);
+  }
+
+  #createFileStream() {
+    if (this.#logFileStream) this.#logFileStream.end();
```
Comment (nechaido) [inline src/logger.js:65]:
```suggestion
    if (this.#logFileStream) {
      this.#logFileStream.end();
    }
```

File: src/logger.js [hunk 12]
```diff
@@ -38,6 +40,37 @@ class Logger {
...
+    }
+    this.#logFileStream.write(msg);
+  }
+
+  #createFileStream() {
+    if (this.#logFileStream) {
+      this.#logFileStream.end();
+    }
+    const filePath = path.join(this.folderPath, `${this.date}.log`);
+    return fs.createWriteStream(filePath, { flags: 'a' });
```
Comment (nechaido) [inline src/logger.js:69]:
```suggestion
    this.#logFileStream = fs.createWriteStream(filePath, { flags: 'a' });
    return this.#logFileStream;
```

File: src/logger.js [hunk 13]
```diff
@@ -38,6 +40,37 @@ class Logger {
...
 
+class StreamForLogger {
+  #logFileStream;
+  folderPath;
+  date;
+
+  constructor(folderPath) {
+    this.folderPath = folderPath;
+    this.date = new Date().toISOString().substring(0, 10); 
+    this.#logFileStream = this.#createFileStream();
```
Comment (nechaido) [inline src/logger.js:51]:
```suggestion
    this.#createFileStream();
```

File: src/logger.js [hunk 14]
```diff
@@ -38,6 +40,38 @@ class Logger {
...
+    this.date = new Date().toISOString().substring(0, 10); 
+    this.#createFileStream();
+  }
+
+  write(msg) {
+    process.stdout.write(msg);
+    const currentDate = new Date().toISOString().substring(0, 10);
+    if (currentDate !== this.date) {
+      this.date = currentDate;
+      this.#logFileStream = this.#createFileStream();
```
Comment (nechaido) [inline src/logger.js:59]:
```suggestion
      this.#createFileStream();
```

**Review by nechaido — APPROVED** [review]
LGTM

**Pull request merged**